pkg/logql/quantile_over_time_sketch.go

```diff
@@ -231,18 +231,36 @@ func (r *quantileSketchBatchRangeVectorIterator) agg(samples []promql.FPoint) sk
 	return s
 }
 
+// Initializing pool
+var quantileVectorPool = sync.Pool{
+	// New optionally specifies a function to generate
+	// a value when Get would otherwise return nil.
+	// New: func() interface{} { return make([]ProbabilisticQuantileVector, 2) },
+	New: func() interface{} { return make([]ProbabilisticQuantileVector, 0, 2) },
+}
+
 // JoinQuantileSketchVector joins the results from stepEvaluator into a ProbabilisticQuantileMatrix.
-func JoinQuantileSketchVector(next bool, r StepResult, stepEvaluator StepEvaluator) (promql_parser.Value, error) {
+func JoinQuantileSketchVector(next bool, r StepResult, stepEvaluator StepEvaluator, params Params) (promql_parser.Value, error) {
 	vec := r.QuantileSketchVec()
 	if stepEvaluator.Error() != nil {
 		return nil, stepEvaluator.Error()
 	}
 
-	result := make([]ProbabilisticQuantileVector, 0)
+	stepCount := int(math.Ceil(float64(params.End().Sub(params.Start()).Nanoseconds()) / float64(params.Step().Nanoseconds())))
+	if stepCount <= 0 {
+		stepCount = 1
+	}
+
+	result := quantileVectorPool.Get().([]ProbabilisticQuantileVector)
+	if cap(result) < stepCount {
+		//  slices.Grow seems to be significantly slower here
+		result = append(make([]ProbabilisticQuantileVector, 0, cap(result)+stepCount), result...)
+	}
+	result = result[:0]
+	defer quantileVectorPool.Put(result)
```

Comment:
Hm, this only works if `ProbabilisticQuantileMatrix(result)` is making a copy which I think it is. @dannykopping what do you think?

Reply:
Hhmm, why do you think it's creating a copy? It just looks like a type-cast here which I don't think creates a copy.

Reply:
It's not making a copy https://go.dev/ref/spec#Conversions
> Specific rules apply to (non-constant) conversions between numeric types or to and from a string type. These conversions may change the representation of x and incur a run-time cost. All other conversions only change the type but not the representation of x.

Reply:
TIL!